src/backend/distributed/executor/repartition_logic.c

```diff
@@ -0,0 +1,252 @@
+	ListCell *taskCell = NULL;
+	List *jobIds = NIL;
+
+	foreach(taskCell, mergeTasks)
+	{
+		Task *task = (Task *) lfirst(taskCell);
+		if (!DoesJobIDExist(jobIds, task->jobId))
+		{
+			jobIds = lappend(jobIds, (void *) task->jobId);
+		}
+	}
+	return jobIds;
+}
+
+
+/*
+ * CreateSchemasOnAllWorkers creates schemas in all workers.
+ */
+static void
+CreateSchemasOnAllWorkers(char *createSchemasCommand)
+{
+	List *commandList = list_make1(createSchemasCommand);
+
+	SendCommandToAllWorkers(commandList);
+}
+
+
+/*
+ * SendCommandToAllWorkers sends the given command to all workers in
+ * a single transaction.
+ */
+static void
+SendCommandToAllWorkers(List *commandList)
+{
+	ListCell *workerNodeCell = NULL;
+	char *extensionOwner = CitusExtensionOwnerName();
+	List *workerNodeList = ActiveReadableNodeList();
+
+	foreach(workerNodeCell, workerNodeList)
+	{
+		WorkerNode *workerNode = (WorkerNode *) lfirst(workerNodeCell);
+		SendCommandListToWorkerInSingleTransaction(workerNode->workerName,
+												   workerNode->workerPort, extensionOwner,
+												   commandList);
+	}
+}
+
+
+/*
+ * GenerateCreateSchemasCommand returns concatanated create schema commands.
+ */
+static char *
+GenerateCreateSchemasCommand(List *jobIds)
+{
+	return GenerateCommand(jobIds, WORKER_CREATE_SCHEMA_QUERY);
+}
+
+
+/*
+ * GenerateCommand returns concatenated commands with the given template
+ * command for each job id from the given job ids. The returned command is
+ * exactly list_length(jobIds) subcommands.
+ *  E.g create_schema(jobId1); create_schema(jobId2); ...
+ * This way we can send the command in just one latency to a worker.
+ */
+static char *
+GenerateCommand(List *jobIds, char *templateCommand)
+{
+	StringInfo createSchemaCommand = makeStringInfo();
+	ListCell *jobIdCell = NULL;
+
+	foreach(jobIdCell, jobIds)
+	{
+		uint64 jobId = (uint64) lfirst(jobIdCell);
+		appendStringInfo(createSchemaCommand, templateCommand, jobId);
+	}
+	return createSchemaCommand->data;
+}
+
+
+/*
+ * DoesJobIDExist returns true if the given job id exists in the given list.
+ */
+static bool
+DoesJobIDExist(List *jobIds, uint64 jobId)
+{
+	ListCell *jobIdCell = NULL;
+	foreach(jobIdCell, jobIds)
+	{
+		uint64 curJobId = (uint64) lfirst(jobIdCell);
+		if (curJobId == jobId)
+		{
+			return true;
+		}
+	}
+	return false;
+}
+
+
+/*
+ * CleanUpSchemas removes all the schemas that start with pg_
+ * in every worker.
+ */
+void
+CleanUpSchemas()
+{
+	List *commandList = list_make1(JOB_SCHEMA_CLEANUP);
+	SendCommandToAllWorkers(commandList);
+}
+
+
+/*
+ * RemoveTempJobDirs removes the temporary job directories that are
+ * used for repartition queries for the given job ids.
+ */
+static void
+RemoveTempJobDirs(List *jobIds)
+{
+	List *commandList = list_make1(GenerateDeleteJobsCommand(jobIds));
+	SendCommandToAllWorkers(commandList);
+}
+
+
+/*
+ * GenerateDeleteJobsCommand returns concatanated remove job dir commands.
+ */
+static char *
+GenerateDeleteJobsCommand(List *jobIds)
```

Comment:
Would be nice to move this next to GenerateCreateSchemasCommand

Reply:
I have placed it after `RemoveTempJobDirs` since it is called in `RemoveTempJobDirs`. I think when someone is looking at `RemoveTempJobDirs` they will want to look at its methods and it is easier to see the methods it calls if they are placed right after it.

We will probably not want to look at `GenerateDeleteJobsCommand` when we are looking at `GenerateCreateSchemasCommand`.

What do you think?